tensorflow/lite/micro/kernels/reduce.cc

```diff
@@ -117,6 +173,86 @@ TfLiteStatus EvalMean(TfLiteContext* context, TfLiteNode* node) {
   }
   return kTfLiteOk;
 }
+
+template <typename T>
+TfLiteStatus EvalLogic(TfLiteContext* context, TfLiteNode* node, T init_value,
+                       T reducer(const T current, const T in)) {
+  const TfLiteTensor* input = GetInput(context, node, 0);
+  const TfLiteTensor* axis = GetInput(context, node, 1);
+  TfLiteTensor* output = GetOutput(context, node, 0);
+  TF_LITE_ENSURE_TYPES_EQ(context, input->type, output->type);
+  TfLiteReducerParams* params =
+      reinterpret_cast<TfLiteReducerParams*>(node->builtin_data);
+  OpData* op_data = static_cast<OpData*>(node->user_data);
+
+  // Interpret an axis tensor with null dimensions as a scalar
+  int num_axis;
+  if (axis->dims == nullptr) {
+    num_axis = 1;
+  } else {
+    num_axis = static_cast<int>(NumElements(axis));
+  }
+
+  int* temp_buffer = static_cast<int*>(
+      context->GetScratchBuffer(context, op_data->temp_buffer_idx));
+  int* resolved_axis = static_cast<int*>(
+      context->GetScratchBuffer(context, op_data->resolved_axis_idx));
+  TF_LITE_ENSURE(
+      context,
+      reference_ops::ReduceGeneric<T>(
+          GetTensorData<T>(input), input->dims->data, input->dims->size,
+          GetTensorData<T>(output), output->dims->data, output->dims->size,
+          GetTensorData<int>(axis), num_axis, params->keep_dims, temp_buffer,
+          resolved_axis, init_value, reducer));
+
+  // Convert between different output scales
+  if (input->type == kTfLiteInt8 &&
+      input->params.scale != output->params.scale) {
+    int8_t* output_data = GetTensorData<int8_t>(output);
+    for (int i = 0; i < NumElements(output); i++) {
+      output_data[i] = static_cast<T>(std::max(
+          std::min(MultiplyByQuantizedMultiplier(
+                       output_data[i], op_data->multiplier, op_data->shift),
+                   static_cast<int>(std::numeric_limits<T>::max())),
+          static_cast<int>(std::numeric_limits<T>::min())));
+    }
+  }
+  return kTfLiteOk;
+}
+
+// Eval for determined input type and reduce type.
+template <typename T>
+TfLiteStatus EvalType(TfLiteContext* context, TfLiteNode* node,
+                      ReduceType reduce_type) {
+  switch (reduce_type) {
+    case kMax:
+      return EvalLogic<T>(context, node, std::numeric_limits<T>::lowest(),
+                          [](const T current, const T in) -> T {
+                            return (in > current) ? in : current;
+                          });
```

Comment:
I understand that this is copied from lite/kernels/reduce.cc but it is a lot of unnecessary complexity for what we currently support in TF Lite Micro.

I would suggest simplifying this code to only support Max reduction and remove all the templates, lambdas and indirection that are currently a distraction.